### PR TITLE
[JENKINS-45233] - Log errors when Response message cannot be delivered due to the closed channel

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Just noticed some issues in command execution logic...

- [x] [JENKINS-45233](https://issues.jenkins-ci.org/browse/JENKINS-45233) - Command processing does not print errors when the Response cannot be delivered due to the closed channel
- [x] Add FINE/FINER logging for better diagnostics of incoming command processing

@reviewbybees

### CODE REVIEW

File: src/main/java/hudson/remoting/Request.java [hunk 1]
```diff
@@ -346,13 +347,17 @@ public void run() {
                         rsp.createdAt.initCause(createdAt);
 
                     synchronized (channel) {// expand the synchronization block of the send() method to a check
-                        if(!channel.isOutClosed())
+                        if(!channel.isOutClosed()) {
                             channel.send(rsp);
+                        } else {
+                            // Propagate exception to the catch clause instead of supressing it
```
Comment (jglick) [inline src/main/java/hudson/remoting/Request.java]: typo

File: src/main/java/hudson/remoting/Request.java [hunk 2]
```diff
@@ -346,13 +347,17 @@ public void run() {
                         rsp.createdAt.initCause(createdAt);
 
                     synchronized (channel) {// expand the synchronization block of the send() method to a check
-                        if(!channel.isOutClosed())
+                        if(!channel.isOutClosed()) {
```
Comment (jglick) [inline src/main/java/hudson/remoting/Request.java]: arguably easier to read if test is inverted

File: src/main/java/hudson/remoting/Request.java [hunk 3]
```diff
@@ -346,13 +347,17 @@ public void run() {
...
                             channel.send(rsp);
+                        } else {
+                            // Propagate exception to the catch clause instead of supressing it
+                            throw new ClosedChannelException();
```
Comment (jglick) [inline src/main/java/hudson/remoting/Request.java]: Is there a known `cause` here?

File: src/main/java/hudson/remoting/Request.java [hunk 4]
```diff
@@ -346,13 +347,17 @@ public void run() {
...
+                        }
                     }
                 } catch (IOException e) {
                     // communication error.
                     // this means the caller will block forever
-                    logger.log(Level.SEVERE, "Failed to send back a reply",e);
+                    logger.log(Level.SEVERE, "Failed to send back a reply to the request " + this, e);
```
Comment (jglick) [inline src/main/java/hudson/remoting/Request.java]: BTW `SEVERE` is probably excessive.

Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Request.java]: Yeah, I would also drop it to WARNING

File: src/main/java/hudson/remoting/Channel.java [hunk 5]
```diff
@@ -528,11 +528,18 @@ protected Channel(@Nonnull ChannelBuilder settings, @Nonnull CommandTransport tr
         transport.setup(this, new CommandReceiver() {
             public void handle(Command cmd) {
                 commandsReceived++;
-                lastCommandReceivedAt = System.currentTimeMillis();
-                if (logger.isLoggable(Level.FINE))
+                long receivedAt = System.currentTimeMillis();
+                lastCommandReceivedAt = receivedAt;
+                if (logger.isLoggable(Level.FINE)) {
                     logger.fine("Received " + cmd);
+                } else if (logger.isLoggable(Level.FINER)) {
```
Comment (jtnord) [inline src/main/java/hudson/remoting/Channel.java]: http://javaadventure.blogspot.co.uk/2016/12/the-costs-of-java-utils-logging.html

File: src/main/java/hudson/remoting/Request.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Request.java]: Actually the check is not required at all. `Channel#send()` does the check

File: src/main/java/hudson/remoting/Channel.java [hunk 6]
```diff
@@ -528,11 +528,18 @@ protected Channel(@Nonnull ChannelBuilder settings, @Nonnull CommandTransport tr
...
+                    logger.log(Level.FINER, "Received command " + cmd, cmd.createdAt);
```
Comment (jtnord) [inline src/main/java/hudson/remoting/Channel.java]: unless cmd.createdAt is an Exception

**Comment by jtnord** [comment]
`Command` does not implement `toString()` - yet it is used in lots of logging - are you intentionally wanting to print the hashCode - in which case I would suggest you replace it's usage with [`System.identityHashCode(it)`](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#identityHashCode-java.lang.Object-)
if you are expecting somethign else then shouldn't you make toString() abstract and force its implementation according to a spec (defined in the classes javadoc)?

File: src/main/java/hudson/remoting/Request.java [hunk 7]
```diff
@@ -346,13 +347,12 @@ public void run() {
                         rsp.createdAt.initCause(createdAt);
 
                     synchronized (channel) {// expand the synchronization block of the send() method to a check
```
Comment (jtnord) [inline src/main/java/hudson/remoting/Request.java]: comment implies this should be removed.

Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Request.java]: right

**Comment by oleg-nenashev** [comment]
> Command does not implement toString() - yet it is used in lots of logging - are you intentionally wanting to print the hashCode - in which case I would suggest you replace it's usage with System.identityHashCode(it)
if you are expecting somethign else then shouldn't you make toString() abstract and force its implementation according to a spec (defined in the classes javadoc)?

Implementations like `UserRequest` and `RPCRequest` have `toString()` implementations, hence I do not consider it as a big problem. But yes, it will be printing class names and hashes for system calls. Not a big deal

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Comment by oleg-nenashev** [comment]
Errors are unrelated

**Pull request merged**